### PR TITLE
[SelectionDAG] Add OPC_MoveSibling

### PR DESCRIPTION
**Pull request by wangpc-pp**
There are a lot of operations to move current node to parent and
then move to another child.

So `OPC_MoveSibling` and its space-optimized forms are added to do
this "move to sibling" operations.

These new operations will be generated when optimizing matcher in
`ContractNodes`. Currently `MoveParent+MoveChild` will be optimized
to `MoveSibling` and sequences `MoveParent+RecordChild+MoveChild`
will be transformed into `MoveSibling+RecordNode`.

Overall this reduces the llc binary size with all in-tree targets by
about 30K.

### CODE REVIEW

File: llvm/utils/TableGen/DAGISelMatcherEmitter.cpp [hunk 1]
```diff
@@ -1066,6 +1077,9 @@ static StringRef getOpcodeString(Matcher::KindTy Kind) {
   case Matcher::RecordMemRef: return "OPC_RecordMemRef"; break;
   case Matcher::CaptureGlueInput: return "OPC_CaptureGlueInput"; break;
   case Matcher::MoveChild: return "OPC_MoveChild"; break;
+  case Matcher::MoveSibling:
+    return "OPC_MoveSibling";
+    break;
```
Comment (topperc) [inline llvm/utils/TableGen/DAGISelMatcherEmitter.cpp:1082]: Can we delete all the unreachable `break`s in this switch as a separate patch?

Comment (wangpc-pp) [reply llvm/utils/TableGen/DAGISelMatcherEmitter.cpp:1082]: Commited as d0c8d41e103dcd3ee1b2b5ed17f9a68e973e01d6.

**Comment by wangpc-pp** [comment]
Ping.

File: llvm/utils/TableGen/DAGISelMatcherOpt.cpp [hunk 2]
```diff
@@ -155,6 +155,31 @@ static void ContractNodes(std::unique_ptr<Matcher> &MatcherPtr,
...
+    if (auto *RC = dyn_cast<RecordChildMatcher>(MP->getNext())) {
+      if (auto *MC = dyn_cast<MoveChildMatcher>(RC->getNext())) {
+        if (RC->getChildNo() == MC->getChildNo()) {
+          auto *MS = new MoveSiblingMatcher(MC->getChildNo());
+          auto *RM = new RecordMatcher(RC->getWhatFor(), RC->getResultNo());
+          // Insert the new node.
+          RM->setNext(MC->takeNext());
+          MS->setNext(RM);
+          MatcherPtr.reset(MS);
+          // MS->setNext(MatcherPtr.release());
```
Comment (arsenm) [inline llvm/utils/TableGen/DAGISelMatcherOpt.cpp:177]: Commented out code

**Review by arsenm — APPROVED** [review]

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**